[Merged by Bors] - Add z-index support with a predictable UI stack

oceantume opened the pull request:
# Objective

Add consistent UI rendering and interaction where deep nodes inside two different hierarchies will never render on top of one-another by default and offer an escape hatch (z-index) for nodes to change their depth.

## The problem with current implementation

The current implementation of UI rendering is broken in that regard, mainly because [it sets the Z value of the `Transform` component based on a "global Z" space](https://github.com/bevyengine/bevy/blob/main/crates/bevy_ui/src/update.rs#L43) shared by all nodes in the UI. This doesn't account for the fact that each node's final `GlobalTransform` value will be relative to its parent. This effectively makes the depth unpredictable when two deep trees are rendered on top of one-another. 

At the moment, it's also up to each part of the UI code to sort all of the UI nodes. The solution that's offered here does the full sorting of UI node entities once and offers the result through a resource so that all systems can use it.

## Solution

### New ZIndex component
This adds a new optional `ZIndex` enum component for nodes which offers two mechanism:
- `ZIndex::Local(i32)`: Overrides the depth of the node relative to its siblings.
- `ZIndex::Global(i32)`: Overrides the depth of the node relative to the UI root. This basically allows any node in the tree to "escape" the parent and be ordered relative to the entire UI.

Note that in the current implementation, omitting `ZIndex` on a node has the same result as adding `ZIndex::Local(0)`. Additionally, the "global" stacking context is essentially a way to add your node to the root stacking context, so using `ZIndex::Local(n)` on a root node (one without parent) will share that space with all nodes using `Index::Global(n)`.

### New UiStack resource
This adds a new `UiStack` resource which is calculated from both hierarchy and `ZIndex` during UI update and contains a vector of all node entities in the UI, ordered by depth (from farthest from camera to closest). This is exposed publicly by the bevy_ui crate with the hope that it can be used for consistent ordering and to reduce the amount of sorting that needs to be done by UI systems (i.e. instead of sorting everything by `global_transform.z` in every system, this array can be iterated over).

### New z_index example
This also adds a new z_index example that showcases the new `ZIndex` component. It's also a good general demo of the new UI stack system, because making this kind of UI was very broken with the old system (e.g. nodes would render on top of each other, not respecting hierarchy or insert order at all).

![image](https://user-images.githubusercontent.com/1060971/189015985-8ea8f989-0e9d-4601-a7e0-4a27a43a53f9.png)

---

## Changelog

- Added the `ZIndex` component to bevy_ui.
- Added the `UiStack` resource to bevy_ui, and added implementation in a new `stack.rs` module.
- Removed the previous Z updating system from bevy_ui, because it was replaced with the above.
- Changed bevy_ui rendering to use UiStack instead of z ordering.
- Changed bevy_ui focus/interaction system to use UiStack instead of z ordering.
- Added a new z_index example.

## ZIndex demo
Here's a demo I wrote to test these features
https://user-images.githubusercontent.com/1060971/188329295-d7beebd6-9aee-43ab-821e-d437df5dbe8a.mp4

oceantume commented:
The local and global z-index solution in this PR is an attempt to bring flexibility without adding too much complexity.

Local z-index allows a node to change its depth compared to its siblings regardless of the order they were added to their parent, while the Global z-index allows nodes to "escape" the parent and be ordered relative to the entire UI.

Those two use-cases are the most common in my opinion, but I can see some benefits in other more complex solutions. For example, having the `ZIndex::Local` depth be relative to the nearest parent node that has a `StackingContext` component attached to it (effectively creating contexts dynamically instead of basing it solely on siblings).

bjorn3 commented:
> This basically allows any node in the tree to

The sentence cut off.

djeedai reviewed (approved):
I really like this PR. I think it not only provides a relatively simple and intuitive solution to a design limitation which currently blocks entire classes of use cases, but it's also a step in what I'd argue is the right direction for UI hierarchies:
- support for multiple UI roots, which this PR helps greatly with its `StackingContext`
- getting rid of the use of `Transform` in UI hierarchies, which unlike other use cases is for UI not user controlled but auto-generated (the source of truth is `Style` for UI), which makes it very unintuitive for users. Unity has the same approach; for UI their Transform is replaced with a RectTransform.

@alice-i-cecile what are the arguments for making this PR as controversial? Can we link a discussion or post some rationale?

TimJentzsch commented on examples/ui/z_index.rs:
```diff
@@ -0,0 +1,223 @@
+//! Demonstrates how to use z-index
+//! Shows two colored buttons with transparent text.
```
Is this correct?

In the video of the PR description there are a lot more elements than two and no transparent text.

oceantume replied:
This is indeed a mistake because the example was improved. I will edit the description.

TimJentzsch reviewed (approved):
Nice, while z-index ordering might not be perfect for all cases, it's definitely a good basis to build on (and familiar to a lot of people).

rparrett commented:
I think that this PR could be improved by

- Addressing how this solution relates to some of the discussion / design from the previous attempt in #1211
- Adding a test. This changeset puts us at -1 tests for "ui z stuff."
- Some benchmarking to demonstrate that there are no major perf regressions.

rparrett commented:
The included example serves as a good demo, but putting myself in the shoes of a user who is trying to learn how `ZIndex` works, it seems pretty difficult to follow.

alice-i-cecile commented:
> Addressing how this solution relates to some of the discussion / design from the previous attempt in https://github.com/bevyengine/bevy/pull/1211

This was what motivated me to add the Controversial label. So long as we can avoid the problems there, I'm happy to move forward on this. I think we have consensus that this needs to happen, and this seems like a good step forward.

Much of the related camera-driven rendering + hierarchy mess that plagued #1211 has since been dramatically improved.

oceantume commented:
> > Addressing how this solution relates to some of the discussion / design from the previous attempt in #1211
> 
> This was what motivated me to add the Controversial label. So long as we can avoid the problems there, I'm happy to move forward on this. I think we have consensus that this needs to happen, and this seems like a good step forward.
> 
> Much of the related camera-driven rendering + hierarchy mess that plagued #1211 has since been dramatically improved.

I made a summary of some important points from that PR and how they're improved upon with this one.

> We need to ensure that layers never "intersect" with each other.

The UiStack concept helps with this since it's a sorted vector built from the hierarchies (and the z-index escape hatch). You can only iterate over it in one way or the other depending on what you need to do.

> I think it probably makes sense to make it relative. Maybe we can do something similar to CSS.

The global and local options helps giving some flexibility here and the result is somewhat similar to CSS with less rules and exceptions.

> I pretty strongly feel that it should be a part of Style (defaulting to a None or Auto value). I don't like requiring a separate component for roots because that is "redundant information".

Correct me if I'm wrong, but it seems bevy_ui is moving away from a single Style struct at this moment so we shouldn't put it in there. The `ZIndex` component is fully optional. However, **we should probably add it to UI bundles** as an `Option<ZIndex>` for clarity/discovery.

> Z precision issues

The z precision issues discussed on there are gone with this solution because we don't move ui nodes on the z axis anymore.

> WindowId, UI to texture concerns

This PR doesn't change anything in that regard.

My PR also doesn't address the fact that multiple root nodes share the same internal taffy root and will share screen space. However, I have a separate PR which does.

oceantume commented:
I just realized that adding `Option<ZIndex>` to the bundles is not yet supported.  Do you think it's important enough that it should be added to the node bundles to justify adding it with a default value of `ZIndex::Local(0)`? It would remain an optional component, but would be present by default on the nodes to help with discovery.

alice-i-cecile commented:
> I just realized that adding Option<ZIndex> to the bundles is not yet supported. Do you think it's important enough that it should be added to the node bundles to justify adding it with a default value of ZIndex::Local(0)? It would remain an optional component, but would be present by default when using bundles to help with discovery.

In this case I think yes. IMO we should prioritize UX over micro-performance pretty heavily for UI.

Remember to hit the other bundles that build off of `NodeBundle` too.

colepoirier commented:
Alice would this close #1275?

oceantume commented:
@rparrett 
> * Some benchmarking to demonstrate that there are no major perf regressions.

Unless I write some new benchmarks, I'm not sure we have any I can run for the UI. I do think they would be beneficial in general, but it's hard for me to benchmark this and compare without doing some full app benchmark that includes a complex nested UI. This moves some z sorting that was done separately in the focus system and render extraction systems into a single, more complex system so that the complexity is only done once. I think we'd need to benchmark a complete app with UI loop to compare the two.

If you have some suggestions for me, please do let me know.

alice-i-cecile commented:
@colepoirier no, because it doesn't work for sprites. It's significant progress though, thanks for linking them.

colepoirier commented:
> @colepoirier no, because it doesn't work for sprites. It's significant progress though, thanks for linking them.

Cool. Is there a fundamental reason that this system can’t be the single system for both ui and sprites for z-order?

oceantume commented:
I have added a unit test for the UI stack system that's similar to the one we had before. I believe this was the last concern for this PR other than for benchmarking, which I've never done with bevy so I wouldn't be sure where to start. Also, I believe there were no benchmarks available for bevy_ui prior to this either.

For me, this looks like ready to go unless other concerns are raised.

oceantume commented:
> > @colepoirier no, because it doesn't work for sprites. It's significant progress though, thanks for linking them.
> 
> Cool. Is there a fundamental reason that this system can’t be the single system for both ui and sprites for z-order?

I think this could work for sprite rendering as well, but it seems to me like it isn't as high of a priority there as it is for UI.

The problem with UI in its current state is that it's simply impossible to have multiple abstract nested trees existing on top of each other. They will intersect in unpredictable ways and there's no fix for it, so I wanted to make something simple that would at least solve that. With sprites, you need to keep track of your Z values, but at least they can be used consistently.

The solutions described in that thread sound like they could be interesting to bring even more flexibility to the system, because my solution is more UI-centric and simplistic.

rparrett commented:
Hovering seems broken right now. No hover-state is shown in the `button` or `many_buttons` examples.

> I believe this was the last concern for this PR other than for benchmarking, which I've never done with bevy so I wouldn't be sure where to start.

I'd probably start by looking at the `many_buttons` example with [tracing / tracy](https://github.com/bevyengine/bevy/blob/main/docs/profiling.md#backend-trace_tracy) and compare execution times of the involved systems.

Here's `ui_focus_system` as an example (Yellow is this PR)

<img width="796" alt="image" src="https://user-images.githubusercontent.com/200550/189037549-f4d91090-7cfa-4de7-b65e-e5aad8736dbf.png">

And `extract_uinodes`:

<img width="796" alt="image" src="https://user-images.githubusercontent.com/200550/189037493-cc8c83cd-f4f7-4c94-9734-da39cddeb355.png">

This is not mega-scientific. I attempted to run my mouse over the window in a similar pattern at a similar speed. Probably something going on there though. At a glance, I'd guess that iterating a list of entities and using `.get` is slower than just iterating a query, which makes sense. Not sure if `iter_many` would help there or not.

alice-i-cecile commented:
As a review note: performance measurements are good, but not blocking for this sort of feature work. Useful, slightly slower libraries are much better than blazing fast ones that are missing essential functionality (like z-order control).

Excellent advice from @rparrett on casual perf investigations though!

alice-i-cecile commented on crates/bevy_ui/src/focus.rs:
```diff
@@ -123,10 +123,14 @@ pub fn ui_focus_system(
+    let moused_over_nodes = ui_stack
+        .uinodes
+        .iter()
+        .rev()
```
Why do we need to reverse here? If we need the list in the opposite order, we should be storing it sorted in the correct order (or using a double-ended queue), rather than constantly flipping it each frame.

djeedai replied:
nit: Agreed on the sentiment, but we're not "flipping it" we're just iterating in reverse order, which I'd assume is only detrimental on very hot loops for memory access locality reasons. However I've asked myself the same question, why are we not storing it in the reverse order if that's the only way we iterate ever (but is it?).

oceantume replied:
The UI focus system needs to traverse the UI nodes from closest to farthest, while the renderer needs to do it from farthest to closest. I'm no expert, but I believe `.rev()` on this iterator is just a reverse loop and it doesn't actually move memory around.

alice-i-cecile replied:
Yep, just verified. Vec iteration won't force reallocation here; VecDeque is needed for double-ended insertion only.

This seems fine to me now; feel free to either leave a comment or just resolve this conversation.

alice-i-cecile commented on crates/bevy_ui/src/ui_node.rs:
```diff
@@ -410,3 +410,24 @@ pub struct CalculatedClip {
     pub clip: Rect,
 }
+
+/// Indicates that this node has special requirements for the order of depth in which
```
```suggestion
/// This [`Node`] entity's front-to-back ordering is not controlled (solely) by its position in the UI hierarchy.
```

alice-i-cecile replied:
(remove line below as well)

djeedai replied:
nit: -position+location? (to avoid confusion with `Transform::position`)

djeedai commented on crates/bevy_ui/src/ui_node.rs:
```diff
@@ -410,3 +410,24 @@ pub struct CalculatedClip {
+pub enum ZIndex {
+    /// Indicates the order in which this node should be rendered relative to its siblings.
+    ///
+    /// For root UI nodes (nodes that have to parent), using this and [`ZIndex::Global`] is equivalent.
```
nit: I'm not sure what "the equivalent" is since we didn't really define equivalence. What about:
```rust
/// For root UI nodes (nodes that have no parent), this is treated as [`ZIndex::Global`].
```

alice-i-cecile replied:
Right, we should specify that it uses the same internal integer.

oceantume replied:
Changed for a more complete description

oceantume commented on crates/bevy_ui/src/focus.rs:
```diff
@@ -123,10 +123,14 @@ pub fn ui_focus_system(
+        .iter()
+        .rev()
+        .filter_map(|entity| {
+            if let Ok((entity, node, global_transform, interaction, _, clip, visibility)) =
```
Can you give me some reference, maybe an example of what we do elsewhere for a somewhat similar case?

alice-i-cecile replied:
The docs here are quite good :D https://docs.rs/bevy/latest/bevy/ecs/query/trait.WorldQuery.html#derive

The basic idea is to create a struct so then we can have named field access to the items.

oceantume replied:
I added a WorldQuery instead. Let me know if this isn't what you had in mind.

rparrett commented on Cargo.toml:
```diff
@@ -1442,6 +1442,16 @@ description = "Demonstrates transparency for UI"
+path = "examples/ui/z_index.rs"
+
+[package.metadata.example.z_index]
+name = "UI Z-INDEX"
```
```suggestion
name = "UI Z-Index"
```

oceantume commented on crates/bevy_ui/src/ui_node.rs:
```diff
@@ -410,3 +410,24 @@ pub struct CalculatedClip {
+/// Indicates that this node has special requirements for the order of depth in which
+/// it should appear in the UI instead of relying solely on the hierarchy of the UI.
+///
+/// Omitting this component on a node will yield the same result as using [`ZIndex::Local(0)`].
```
Hopefully it's much clearer now.

oceantume commented:
I've made a few changes:
- Improvements to the docs for clarity
- Added a `WorldQuery` in the focus system to improve readability
- Changed the focus and render extraction systems to use `iter_many` for readability (and possibly performances)

rparrett commented:
That looks a lot nicer at a glance.

I am still seeing broken "hovering" behavior in the button example (pre-iter-mut as well). I believe that there was something subtle going on in the previous implementation because now the
```
    // reset lower nodes to None
```
code is immediately resetting the nodes marked as `Hovered` in the code above.

alice-i-cecile commented on crates/bevy_ui/src/focus.rs:
```diff
@@ -62,6 +62,19 @@ pub struct State {
     entities_to_reset: SmallVec<[Entity; 1]>,
 }
 
+/// Main query for the focus system
```
```suggestion
/// Main query for [`ui_focus_system`]
```

alice-i-cecile commented on crates/bevy_ui/src/focus.rs:
```diff
@@ -161,52 +168,49 @@ pub fn ui_focus_system(
         }
 
-        match focus_policy.cloned().unwrap_or(FocusPolicy::Block) {
+        match node.focus_policy.cloned().unwrap_or(FocusPolicy::Block) {
```
Can we just make `FocusPolicy` a `Copy` type instead?

oceantume replied:
It's already copy. `cloned` was used here to transform `Option<&FocusPolicy>` to `Option<FocusPolicy>`. I went the other way around and simplified with `node.focus_policy.unwrap_or(&FocusPolicy::Block)`.

alice-i-cecile reviewed (changes requested):
I really like what `NodeQuery` did to the code :D

Some small items of feedback, but the introduced bug is my only blocking concern.

rparrett commented:
`iter_many` is actually perhaps salvageable, but I haven't benchmarked yet to determine if it's worth the extra complexity.

https://github.com/rparrett/bevy/commit/ffbe0dcae7c9c073b25dc363ae42c08625283151

oceantume commented:
> `iter_many` is actually perhaps salvageable, but I haven't benchmarked yet to determine if it's worth the extra complexity.
> 
> [rparrett@ffbe0dc](https://github.com/rparrett/bevy/commit/ffbe0dcae7c9c073b25dc363ae42c08625283151)

It seems to me like performances for this will probably be at best similar (but possibly worse) than just using `.get_mut(entity)` in a loop? You're now going through the iterator item by item to find the right one on every loop, whereas the `get_mut(entity)` option at least relies on bevy's lookup performance.

I implemented `iter_mut` quickly, assuming that it would give me results in the order that they are in the provided vector, but if it doesn't do that by design then I guess there's not much we can do other than revert this part or implement an `iter_mut` alternative that actually does that.

oceantume commented:
Indeed, I just realized my mistake while looking at the code and trying an alternative. I looked at the internal code for `.iter_many()` and it essentially does a `.get()` on every entity in the loop, so my understanding is that there won't be much benefit to using `iter_many`.

What do you think of this which keeps the enumerate order? It's a bit verbose but I think it may be easier to understand what's going on. Note that the UI focus system can just keep using `iter_many_mut` since it doesn't need to keep the indices.
```
    for (stack_index, (uinode, global_transform, text, text_layout_info, visibility, clip)) in
        ui_stack
            .uinodes
            .iter()
            .enumerate()
            .filter_map(|(stack_index, entity)| {
                uinode_query
                    .get(*entity)
                    .map(|result| (stack_index, result))
                    .ok()
            })
    {
      // ...
    }
```

djeedai commented:
I'd like to go back to @alice-i-cecile comment :

> performance measurements are good, but not blocking for this sort of feature work.

Can we move ahead with this PR as is, and leave any potential optimization for a later pass? The discussion is very interesting, but there's already a lot of work here with clear value over the existing solution, I wouldn't want to discourage @oceantume nor risk the work getting stuck in PR limbo. I'd prefer we focus on the last items if any (I didn't follow, is there still a bug to fix?) and merge what we have, which fixes a major blocker for UI hierarchies. Thanks ❤️

rparrett commented:
"as-is" is currently completely broken, so that needs to be resolved.

But yeah if I haven't been clear, my position is / has been that we are chasing microseconds of frame time (but we didn't know that until we measured). Please carry on with `.get`.

oceantume commented:
I reverted the render extraction system to the unbroken state it was in before using `iter_many`, fixed the hover system and resolved the remaining comments.

oceantume commented:
FYI this is ready for final review for me.

I think I don't understand GH's review system because I keep removing someone's review request when I request a review from someone else.

rparrett reviewed (approved):
Things seem to be working, and perf seems comparable. If there's consensus that this is "uncontroversial," then I am happy.

alice-i-cecile commented on crates/bevy_ui/src/focus.rs:
```diff
@@ -172,41 +179,43 @@ pub fn ui_focus_system(
     // reset lower nodes to None
-    for (_entity, _focus_policy, interaction, _) in moused_over_z_sorted_nodes {
-        if let Some(mut interaction) = interaction {
+    let mut iter = node_query.iter_many_mut(moused_over_nodes);
```
This pattern needs a comment to explain why we're using it  + the underlying logic.

oceantume replied:
Would you prefer if it used a simpler loop with `get_mut` instead? I changed this with the hope that it may improve performances, but it's unclear if it really does and maybe clarity would be better here? Otherwise I can add a simple comment, something along the line of:
```
// use the [`iter_many_mut`] pattern to go over remaining nodes. it allows iterating over a specific
// subset of entities in the node query but it doesn't support using it as a simple [`Iterator`].
```

oceantume replied:
You can ignore my previous question. I added multiple new comments in the code to explain what's happening and why.

alice-i-cecile replied:
LGTM now!

alice-i-cecile commented on examples/ui/z_index.rs:
```diff
@@ -0,0 +1,126 @@
+//! Demonstrates how to use z-index
```
```suggestion
//! Demonstrates how to use the z-index component on UI nodes to control their relative depth
```

alice-i-cecile commented on examples/README.md:
```diff
@@ -314,6 +314,7 @@ Example | Description
 [Transparency UI](../examples/ui/transparency_ui.rs) | Demonstrates transparency for UI
 [UI](../examples/ui/ui.rs) | Illustrates various features of Bevy UI
 [UI Scaling](../examples/ui/scaling.rs) | Illustrates how to scale the UI
+[UI Z-Index](../examples/ui/z_index.rs) | Demonstrates z-index for UI
```
```suggestion
[UI Z-Index](../examples/ui/z_index.rs) | Demonstrates how to control the relative depth (z-position) of UI elements
```

alice-i-cecile replied:
When writing explanations, don't try to explain concepts in terms of themselves :)

alice-i-cecile commented on crates/bevy_ui/src/ui_node.rs:
```diff
@@ -410,3 +410,31 @@ pub struct CalculatedClip {
+/// (nodes that have no parent). Because of this, there is no difference between using
+/// [`ZIndex::Local(n)`] and [`ZIndex::Global(n)`] for root nodes.
+///
+/// Not inserting this component on a node will give the same result as using [`ZIndex::Local(0)`].
```
```suggestion
/// Nodes without this component will be treated as if they had a value of [`ZIndex::Local(0)`].
```

oceantume commented:
Updated documentation to resolve comments!

alice-i-cecile reviewed (approved):
I'm happy with this now. @cart, what's your ruling on controversial status? It's foundational and relatively large, but we have strong consensus from the team and no serious alternative has been proposed.

alice-i-cecile commented:
bors try

cart commented:
@alice-i-cecile I would still call this controversial as this is a pretty foundational piece of UI that will be hard to change once people start depending on its behaviors. I'd like to give this a look, but I also don't want to hold back progress. Lets "start the merge clock" on this, but with the agreement that it can be merged for Bevy 0.9 if I don't get to it in time (so _slightly_ accelerated).

alice-i-cecile commented:
Sounds good!

cart commented on crates/bevy_ui/src/render/mod.rs:
```diff
@@ -382,10 +392,10 @@ pub fn prepare_uinodes(
     extracted_uinodes
         .uinodes
-        .sort_by(|a, b| FloatOrd(a.transform.w_axis[2]).cmp(&FloatOrd(b.transform.w_axis[2])));
+        .sort_by_key(|node| node.stack_index);
```
Its a bit unfortunate that we already have an absolute order in UiStack, which we're then re-establishing here in uinodes. The sort is redundant. That being said, if we want a "distributed" extract (ex: multiple systems with node-type-specific extraction logic), then the only real alternative is extracting nodes as entities (instead of pushing them onto an unsorted ExtractedUiNodes list), then using an extracted UiStack value to query/iterate over them in `prepare_uinodes`. The extra cost of "per-entity extraction" might outweigh the time saved on sorting.

Worth measuring, but not something we need to block on here, especially given that this is following the pre-existing pattern.

cart reviewed (commented):
I'm on board for this! The sorting algorithm is usable and roughly in line with the web spec. There may be potential here for optimization in various places, but I'm cool with moving forward on this as-is (once merge conflicts are resolved).

alice-i-cecile reviewed (approved):
Merge looks correct.

oceantume commented on crates/bevy_ui/src/stack.rs:
```diff
@@ -0,0 +1,222 @@
+    }
+
+    *ui_stack = UiStack {
+        uinodes: Vec::<Entity>::with_capacity(total_entry_count),
```
I just made changes to do that. Would just need a confirmation that it makes sense as-is.

james7132 commented:
@oceantume does this address #1275?

alice-i-cecile commented:
IMO no, not fully. It doesn't work for 2D, and doesn't allow a direct and automatic translation from an ordered enum.

cart commented:
Just calling out that the recent merge broke this code / it needs fixing.

cart commented:
bors r+

oceantume commented:
Thanks for fixing it @cart. I was hoping to look into this today, but didn't find the time.
